### ISSUE TITLE
minSize, maxSize not functioning on window resize

### ISSUE THREAD

**Issue by Luxapodular** [leefsmp/Re-Flex#62]
Hi!

On window resize, the minSize and maxSize restrictions do not seem to be functioning properly. This seems to be due to the fact that the "flex" style properties are not being readjusted as the window changes size. 

Here's a gif of a div breaking both a min and max size property.
![reflex resize problem](https://user-images.githubusercontent.com/8699557/47680902-80fa4080-db9e-11e8-9358-b496a0e37aad.gif)

A subtle hack is to add a min-width, min-height, max-width, or max-height property to the div which should be bounded. However, this causes the reflex splitters to lag behind the actual size of the div if the window is resized.

React-reflex has been very helpful for our project, so thank you very much for the work so far. I'd love to give a hand fixing this issue / testing it as soon as possible.

**Comment by Luxapodular** [comment]
Here's what the reflexContainer and elements look like in the above example!

```
  <ReflexContainer orientation="vertical">
    <ReflexElement className="center-container">
    </ReflexElement>
    <ReflexSplitter />
    <ReflexElement
      className="right-container"
      size={350} minSize={200}
      maxSize={350}
      renderOnResize={true}
      propagateDimensions={true}>
    </ReflexElement>
  </ReflexContainer>
```

**Comment by leefsmp** [comment]
Hi Luca,

I think the reflex logic is not being recomputed automatically when user manually resize the browser window, one way to trigger that would be to add a window resize event which forces the update of the   component containing the layout:

   window.addEventListener(
      'resize', this.onResize)

    onResize = () => {
        this.forceUpdate()
    }

You may have to add custom logic based on the actual size of the window that makes sure the min/max contraints passed to `Reflex` are achievable, for example setting a `minSize={500}` when the window is less than that will be a problem. Let me know if that helps.

**Comment by Luxapodular** [comment]
Actually, after testing a bit more this does have 1 relatively bad, but potentially avoidable side effect. Rerendering the ReflexContainer will cause the elements within the container to jump back to default size. 

![jumpback](https://user-images.githubusercontent.com/8699557/47730815-527d7380-dc39-11e8-9a34-9e85e4100f8b.gif)

This could be avoided by using local storage to store the size of the component on resize, similarly to the provided example. However, I can't seem to figure out how to pull size information from the components in pixels rather than their "flex" property. As mentioned above, the flex property will be invalid on a window resize. 

Any thoughts on this issue as well?

**Comment by leefsmp** [comment]
Its a bit of a tricky trade of... yes rerendering the container will reset the internal state and therefore resize the elements to the size defined by their props (+ applied constraints if any). 

I see two ways you could deal with this: 

1# use the `propagateDimensions={true}` on `ReflexElement` like in the [Size-Aware element demo](https://leefsmp.github.io/Re-Flex/index.html#demo7). But doing this will trigger a rerender of the elements at each resize step, so depending how heavy are your elements to render this might not be desirable. You may want to optimize each element content using `shouldComponentUpdate` for example.

2# use the element events to get a notification  when the user is resizing a layout element. [Here is a pen](https://codepen.io/leefsmp/pen/YJboqN) that shows how to compute the size in pixel of each resized elements.

But setting the size in pixel of your elements upon resizing somehow defeat the purpose of using a flex layout. I think the best in that scenario would be to adapt your min/max constraints based on the size of the window upon resize but setting minSize/maxSize on an element after the first initialization of the layout will not recompute the internal state in the current version of the lib. I may need to think a bit more about your case...

**Comment by Luxapodular** [comment]
Hi @leefsmp, Apologies for the delayed response! 

I've tried the proposed solutions, and neither of them seem to work. 

#1 Using propagateDimensions seems to not cause the element to re render on window resize. While the ReflexElements within the element which is propagating dimensions are re rendering, the parent reflex element does not recalculate its bounds!

#2 While this can be used by hooking this function into windowResize, it seems that natively the ReflexElement will not call it's resize functions on window resize. Very odd. Maybe I'm doing something wrong here.

The 'key' update method still prevails here though.

**Comment by Luxapodular** [comment]
Also, this may be a separate issue, but it seems that ReflexElements are forcing their children to update without propagateDimensions={true} being set. 

For instance in this example elemA is rerendered every single time the Reflex Element changes size, but elemB is not.
 
```
<ReflexElement>
  <elemA>
    <elemB>
    </elemB>
  </elemA>
</ReflexElement>
```

**Comment by leefsmp** [comment]
A simple solution for that is to make `elemA` a `PureComponent` l[ike in this pen](https://codepen.io/leefsmp/pen/WYeMgx) if you are using React 16+ or implement a custom `componentShouldUpdate` if you are using a lower React version.

**Comment by leefsmp** [comment]
Getting back to the original topic, I think the best way to handle that scenario would be to set an event listener for window resize event internal to the `ReflexContainer` which will trigger a recompute of the layout based on the current internal flex values, which will prevent the elements to resize to their default size, however it seems tricky to find what value is appropriate when computing the flex of each elements based on the user constraints (minSize/maxSize/size) because the system needs some degrees of freedom in order to be able to fit elements in the layout. I am working on a `v3.0.0` which I am planning to release over the next couple of weeks, so I will see what can be done about it.

In the meantime, [the following pen](https://codepen.io/leefsmp/pen/aQoKPK) is addressing this behavior in a specific case: sets up `window.resize` event and triggers update of the layout using new `key` prop as you mentioned earlier. When right pane is being user-resized, keep track of the `flex` value and pass it as prop so it becomes a constraint and the layout is not reset to default.

**Comment by leefsmp** [comment]
I've added a `resizeWindowAware` property on `ReflexContainer` in `v.3.02`. This suppress the need to use the `key` hack discussed above to force a rerender but I couldn't find an acceptable way to maintain the size constraint automatically on the resized panel which would be built in the lib. 

It seems acceptable to me to take that kind of customization of the consumer side by adding some extra logic as illustrated in this [pen](https://codepen.io/leefsmp/pen/WYbvbm). Closing the ticket for now unless you have further comments.
